md5sum.hpp

```diff
@@ -0,0 +1,83 @@
+// Compute checksums of files or strings.
+//
+// Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019 Gregory W. Chicares.
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 2 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software Foundation,
+// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+//
+// http://savannah.nongnu.org/projects/lmi
+// email: <gchicares@sbcglobal.net>
+// snail: Chicares, 186 Belle Woods Drive, Glastonbury CT 06033, USA
+
+#ifndef md5sum_hpp
+#define md5sum_hpp
+
+#include "config.hpp"
+#include "so_attributes.hpp"
+
+#include <climits>                      // CHAR_BIT
+#include <string>
+#include <vector>
+
+// The gnu libc md5 implementation seems to assume this:
+static_assert(8 == CHAR_BIT || 16 == CHAR_BIT);
+// so md5 output is 128 bits == 16 8-bit bytes or 8 16-bit bytes:
+enum { md5len = 128 / CHAR_BIT };
+
+enum { chars_per_formatted_hex_byte = CHAR_BIT / 4 };
+
+struct md5sum_for_file
+{
+    md5sum_for_file(std::string file, std::string sum)
+        :filename{file}
+        ,md5sum(sum)
+        {
+        }
+
+    std::string filename;
+    std::string md5sum;
+};
+
+/// Reads the vector of structs with the file name and the md5 sum from the given file.
+///
+/// The input stream contains hash and filename pairs.
+/// Sample:
+/// 595f44fec1e92a71d3e9e77456ba80d1  filetohashA.txt
+/// 71f920fa275127a7b60fa4d4d41432a3  filetohashB.txt
+/// 43c191bf6d6c3f263a8cd0efd4a058ab  filetohashC.txt
+///
+/// There must be two spaces or a space and an asterisk between each md5sum
+/// value and filename to be compared (the second space indicates text mode,
+/// the asterisk binary mode). Otherwise, the following error will result:
+/// "no properly formatted MD5 checksum lines found".
+/// The current implementation don't distinguish between the two modes.
+
+bool LMI_SO md5sum_vec_from_file
```

Comment:
Not sure if this convention is already used in lmi, but I'd rather return the error message as string instead of bool, with the convention that the function succeeds only if the returned string is empty. This avoids the implicit dependency between the return value and `error_message` output argument, i.e. it makes it impossible to return `true` but not fill the error message or, on the contrary, return `false` while filling it.